lib/streamlit/testing/v1/element_tree.py

```diff
@@ -0,0 +1,1506 @@
+    Step,
+)
+from streamlit.elements.widgets.time_widgets import (
+    DateInputSerde,
+    DateWidgetReturn,
+    TimeInputSerde,
+    _parse_date_value,
+)
+from streamlit.proto.Block_pb2 import Block as BlockProto
+from streamlit.proto.Button_pb2 import Button as ButtonProto
+from streamlit.proto.Checkbox_pb2 import Checkbox as CheckboxProto
+from streamlit.proto.Code_pb2 import Code as CodeProto
+from streamlit.proto.ColorPicker_pb2 import ColorPicker as ColorPickerProto
+from streamlit.proto.DateInput_pb2 import DateInput as DateInputProto
+from streamlit.proto.Element_pb2 import Element as ElementProto
+from streamlit.proto.Exception_pb2 import Exception as ExceptionProto
+from streamlit.proto.ForwardMsg_pb2 import ForwardMsg
+from streamlit.proto.Heading_pb2 import Heading as HeadingProto
+from streamlit.proto.Markdown_pb2 import Markdown as MarkdownProto
+from streamlit.proto.MultiSelect_pb2 import MultiSelect as MultiSelectProto
+from streamlit.proto.NumberInput_pb2 import NumberInput as NumberInputProto
+from streamlit.proto.Radio_pb2 import Radio as RadioProto
+from streamlit.proto.Selectbox_pb2 import Selectbox as SelectboxProto
+from streamlit.proto.Slider_pb2 import Slider as SliderProto
+from streamlit.proto.Text_pb2 import Text as TextProto
+from streamlit.proto.TextArea_pb2 import TextArea as TextAreaProto
+from streamlit.proto.TextInput_pb2 import TextInput as TextInputProto
+from streamlit.proto.TimeInput_pb2 import TimeInput as TimeInputProto
+from streamlit.proto.WidgetStates_pb2 import WidgetState, WidgetStates
+from streamlit.runtime.state.common import user_key_from_widget_id
+from streamlit.runtime.state.session_state import SessionState
+
+if TYPE_CHECKING:
+    from streamlit.testing.v1.app_test import AppTest
+
+T = TypeVar("T")
+
+
+@dataclass
+class InitialValue:
+    """This class is used to represent the initial value of a widget."""
+
+    pass
+
+
+# TODO This class serves as a fallback option for elements that have not
+# been implemented yet, as well as providing implementations of some
+# trivial methods. It may have significantly reduced scope, or be removed
+# entirely, once all elements have been implemented.
+# This class will not be sufficient implementation for most elements.
+# Widgets need their own classes to translate interactions into the appropriate
+# WidgetState and provide higher level interaction interfaces, and other elements
+# have enough variation in how to get their values that most will need their
+# own classes too.
+@dataclass
+class Element:
+    type: str
+    proto: Any = field(repr=False)
+    root: ElementTree = field(repr=False)
+    key: str | None
+
+    def __init__(self, proto: ElementProto, root: ElementTree):
+        self.proto = proto
+        self.root = root
+        ty = proto.WhichOneof("type")
+        assert ty is not None
+        self.type = ty
+        self.key = None
+
+    def __iter__(self):
+        yield self
+
+    @property
+    def value(self) -> Any:
```

Comment:
Wouldn't it be possible to move `value` and `widget_state` to `Widget`?

Reply:
Non-widget elements also have values. `widget_state` is present on everything so that `ElementTree.get_widget_states()` could handle all node types uniformly. I initially wrote it this way because the nodes are very heterogenous so it felt natural. I normally prefer only defining methods that make actual sense for the specific type, and this might be better written using a separate function that dispatches based on the node's type, rather than relying on subtype polymorphism for dispatch.